test/checkpoint-writer.spec.js

```diff
@@ -0,0 +1,102 @@
+'use strict'
+const expect = require('chai').expect;
+const Joi = require('joi');
+const sinon = require('sinon');
+const checkpointWriter = require('../lib/events-reader-checkpoint-writer');
+const Promise = require('bluebird');
+
+describe('checkpoint writer', function () {
+
+  describe('timeout', function () {
+    context('when using the default config options', () => {
+      const harvestApp = {
+        adapter: {
+          update: () => {}
+        }
+      };
+      const fakeDoc = {ts: 1};
+      const checkpointEvent = checkpointWriter.checkpointEvent;
+
+      let clock;
+
+      beforeEach(() => {
+        clock = sinon.useFakeTimers();
+        sinon.stub(harvestApp.adapter, 'update');
+        harvestApp.adapter.update.returns(new Promise.resolve(fakeDoc));
+        checkpointWriter.startWriterLoop(harvestApp);
+        checkpointWriter.setWriterLoopStopped(true);
+        checkpointEvent.emit('newCheckpoint', 1, fakeDoc);
+      });
+
+      afterEach(() => {
+        harvestApp.adapter.update.restore();
+        clock.restore();
+      });
+
+      it('should clean last doc and checkpoint after handled', done => {
+        clock.tick(1);
+        expect(harvestApp.adapter.update.callCount).to.be.eql(1);
+        clock.tick(1);
+        expect(checkpointWriter.getLastDoc()).to.be.undefined;
+        expect(checkpointWriter.getLastCheckpointId()).to.be.undefined;
+        expect(harvestApp.adapter.update.calledOnce).to.be.true;
+
+        done();
+      });
+
+      it('should write a checkpoint in a given interval', done => {
+        clock.tick(1);
+        expect(harvestApp.adapter.update.callCount).to.be.eql(1);
+
+        checkpointEvent.emit('newCheckpoint', 1, fakeDoc);
+        clock.tick(1);
+        expect(harvestApp.adapter.update.callCount).to.be.eql(2);
+
+        done();
+      });
+    });
+
+    context('when passing the option eventsReaderDebounceWait', () => {
+      const eventsReaderDebounceDelay = 1000;
+      const harvestApp = {
+        adapter: {
+          update: () => {}
+        },
+        options: {
+          eventsReaderDebounceWait: eventsReaderDebounceDelay
+        }
+      };
+      const fakeDoc = {ts: 1};
+      const checkpointEvent = checkpointWriter.checkpointEvent;
+
+      let clock;
+
+      beforeEach(() => {
+        clock = sinon.useFakeTimers();
+        sinon.stub(harvestApp.adapter, 'update');
+        harvestApp.adapter.update.returns(new Promise.resolve(fakeDoc));
+        checkpointWriter.startWriterLoop(harvestApp);
+        checkpointWriter.setWriterLoopStopped(true);
+        checkpointEvent.emit('newCheckpoint', 1, fakeDoc);
+      });
+
+      afterEach(() => {
+        harvestApp.adapter.update.restore();
+        clock.restore();
+      });
+
+      it('should write a checkpoint in a given interval', done => {
```

Comment:
This seems identical to line 47; i dont know that you're testing very much here. You might want to check that it debounces writes, so you call 500 writes in 1ms and expect 1 write.